### PR TITLE
Add support for IO-blocks

### PR DESCRIPTION
**Pull request by sergiorg-hpc**
The improvements provided in #174 and available in `libsonata v0.1.11`, introduced a specific limitation while fetching a small number of widely-disperse GIDs in very large reports. The `{min,max}` optimization, inspired by [ROMIO's Data Sieving](https://web.cels.anl.gov/~thakur/papers/romio-coll.pdf), would in such case force the implementation to retrieve large amounts of data without considering the gap between each of the position ranges of the GIDs. This means that, if we fetch only 2 GIDs that coincidentally are located in the very first entry of the dataset and in the very last one, the library would then read the complete report file just to filter in memory the two selected values.

To overcome these constraints while still trying to reduce the number of IO requests requested to the parallel file system, the current PR introduces several changes and optimizations:

- The `{min,max}` range is now replaced by a collection of ranges or IO blocks, in which each range contains a subset of GIDs. The gap between each IO block is fixed at 64MB (or 4 blocks in GPFS), allowing the code to adapt depending on the number of GIDs requested and their distribution in the file. For instance, a higher number of GIDs would generally imply a lower number of IO blocks, and vice versa.
- A flat-map inspired implementation replaces the existing `std::map` that is utilized to index the GIDs and their positions. The idea is to reduce the overhead when opening populations from very large report files with millions of GIDs. Moreover, it allows us to re-index the content based on different keys, without moving the actual data. For this purpose, the implementation uses several `std::vector` and a dedicated index that refers to the values contained in these vectors. When creating the IO blocks, it is enough to re-index the content by the position ranges and to create the groups based on this new index.
- Other minor changes are included, such as behaviour consistency between asking for all the GIDs or a subset of the GIDs, clarification comments, and others.

Using the same ~60TB report utilized in #174 and #179, the following table illustrates the performance difference between `v0.1.11` and the proposed PR by fetching a single timestep while varying the number of GIDs requested. The `Open` column reflects the cost of opening the population as part of the elapsed execution time, while the `Select` column shows the cost of the operations related to the selection of the GIDs in Python, and the `Get` column illustrates the cost of the actual retrieval of the data. The total cost is reflected in `Elapsed`. All measurements are provided in seconds:

```
v0.1.11
-------------------------------------------------------
   GIDs |   Elapsed |      Open |    Select |       Get
      1 |  5.347993 |  4.720139 |  0.626175 |  0.001667
     10 | 11.422943 |  5.391727 |  0.781538 |  5.249666
    100 | 11.814635 |  5.087994 |  0.763447 |  5.963183
   1000 | 12.090420 |  5.105253 |  0.774254 |  6.210903
  10000 | 12.084519 |  5.047278 |  0.779182 |  6.258047
 100000 | 12.789811 |  5.165905 |  0.785131 |  6.838766
1000000 | 18.066422 |  5.148455 |  0.834916 | 12.083038
4234929 | 33.317796 |  5.097557 |  0.414970 | 27.805258

PR
-------------------------------------------------------
   GIDs |   Elapsed |      Open |    Select |       Get
      1 |  1.558976 |  0.933911 |  0.623337 |  0.001715
     10 |  1.450080 |  0.733224 |  0.628178 |  0.088669
    100 |  4.454307 |  0.749991 |  0.638895 |  3.065413
   1000 |  7.777399 |  0.803403 |  0.632553 |  6.341434
  10000 |  7.836496 |  0.821947 |  0.635333 |  6.379207
 100000 |  8.150157 |  0.815197 |  0.636635 |  6.698314
1000000 | 12.450882 |  0.806348 |  0.706298 | 10.938223
4234929 | 21.047557 |  0.780823 |  0.272458 | 19.994266
```

From the results, it can be observed that the use of multiple-blocks and the integration of a flat-map reduces the overhead of the execution time considerably, specially in small number of GIDs. After 100 GIDs, the implementation can only find a single IO block to fetch from storage and, thus, its behaviour is equivalent to the original `{min,max}` single-block implementation of `v0.1.11`. Hence, allowing us to be more flexible in lower number of GIDs, and more aggressive in large number of GIDs to prevent overheads in the parallel file system. This is clearly noticeable by observing the `Get` column of the table above (note that the performance is improved because the PR also contains the changes in #179).

The following table reflects the number of blocks created by the implementation for each of the tests, showing how we can now be more flexible in requests with lower number of GIDs and more aggressive in requests with large number of GIDs:

```
PR
-----------------
   GIDs |  Blocks
      1 |       1
      2 |       2
     10 |      10
    100 |      32
   1000 |       1
  10000 |       1
 100000 |       1
1000000 |       1
4234929 |       1
```

If we now evaluate the edge case mentioned in the beginning of this description, the entry with 2 GIDs in the following table illustrates the performance by fetching the very first and the very last GID of the same ~60TB report for a single timestep:

```
v0.1.11
-------------------------------------------------------
   GIDs |   Elapsed |      Open |    Select |       Get
      2 | 11.180127 |  4.742555 |  0.205418 |  6.232141

PR
-------------------------------------------------------
   GIDs |   Elapsed |      Open |    Select |       Get
      2 |  1.103376 |  0.893741 |  0.209070 |  0.000555
```

### CODE REVIEW

**Comment by sergiorg-hpc** [comment]
Note that I'll update this PR after #182 is merged. Under my opinion, the logic plan would be to merge first #182, then try to merge #183 after it is ready, and finally refactor the code as proposed in #181.

File: src/report_reader.cpp
Comment (matz-e) [inline src/report_reader.cpp:350]: In theory, we could add some parallel execution policy here, if desired.

**Review by matz-e — COMMENTED** [review]
Had an initial look over this, looks good to me!

Thank you as always for your work!

**Comment by mgeplf** [comment]
Nice speedup!

> Under my opinion, the logic plan would be to ...

That makes sense; don't worry about #181, I can just redo that when I have time.  I just merged #182, so we should be good to go.

File: src/report_reader.cpp
Comment (mgeplf) [reply src/report_reader.cpp:350]: IMO, I think we should keep this change simple, and potentially do a pass at some point in the future to change execution policies in bulk.

File: .gitignore
Comment (mgeplf) [inline .gitignore:4]: good call on the ignores

File: src/report_reader.cpp
Comment (mgeplf) [reply src/report_reader.cpp:345]: I agree that it would be a good idea; `getenv` is a funny function to deal with, the threadsafety is sus, and iirc, it triggers locale work (prob. only on first run?)  I try to avoid it, mainly due to previously being burned by it.

Comment (matz-e) [reply src/report_reader.cpp:345]: Is there an alternative to `getenv`? I think I'd feel better if this was done statically at initialization time.

File: src/report_reader.cpp
Comment (sergiorg-hpc) [reply src/report_reader.cpp:327]: The `std::binary_search` returns a `bool`, so I use the `std::lower_bound` to get the actual iterator through a binary search. If there are multiple GIDs that are actually the same, the code would still work (compared to the old implementation, that was unifying duplicates into one through the `std::map`).

File: src/report_reader.cpp
Comment (sergiorg-hpc) [reply src/report_reader.cpp]: Thank you guys for the comments. I have replaced the `getenv` with an extra parameter to the function and a boundary check, to prevent the users to set very small gaps and potentially force the code to trigger a hyperslab per GID (as in the original code changed in #174). The Python bindings are also updated and there are now CI tests for this as well.

**Comment by sergiorg-hpc** [comment]
A quick update. I have adapted the code of the PR after rebasing from `master`, including the use of `std::next` and `std::advance` with iterators instead of raw pointers. I have also replaced the ENV variable for the `block_gap_size` with a parameter, that is also exposed through Python. The parameter includes a boundary check to prevent users affecting the performance on GPFS and there are new CI tests to guarantee that we do not change this accidentally in the future. Finally, I have also verified that the performance remains equivalent to what we had with raw pointers and the use of the ENV variable.

I will wait for further comments from your side. Also, thanks to @matz-e and @iomaganaris for the insight.

File: src/report_reader.cpp
Comment (sergiorg-hpc) [reply src/report_reader.cpp:244]: Excuse me, I added it but perhaps I reverted the change at some point without noticing. I have just pushed the change back to how it was with the `SonataError` in-place.

File: tests/test_report_reader.cpp
Comment (mgeplf) [inline tests/test_report_reader.cpp:92]: we weren't checking the data output here? I'll add a test for that (#187), so that we can check the `data_all.data` is moving around as expected.  Once that's in, can you merge and update it as required?

**Review by mgeplf — COMMENTED** [review]
looks good, I'd just like #187 to be merged first, so that the change in order of the report reading is checked.

File: include/bbp/sonata/report_reader.h
Comment (sergiorg-hpc) [reply include/bbp/sonata/report_reader.h:187]: Exactly, I found it surprising dangling there with no use. I double-checked the attribute with @jorblancoa and he confirmed that it does exist.

File: src/report_reader.cpp
Comment (sergiorg-hpc) [reply src/report_reader.cpp:85]: We used to have something like `result.ids.emplace_back(make_key<T>(...))` here, because the vector in the result was not resized but "reserved". Now the code needs random access to the output array due to the IO-blocks and we can no longer grow the vector element by element. Thus, we resize the vector first and then copy the values to the actual position according to the positions indexed and stored internally.

I'm not sure how to solve this differently, because `T` can either be a `uint64_t` or a `std::array<uint64_t, 2>`. What would be your approach on this? @matz-e suggested `constexpr`, but it is only available in C++17 and above.

File: src/report_reader.cpp
Comment (sergiorg-hpc) [reply src/report_reader.cpp:321]: No, not exactly. Each GPFS block is 16MB, so having a limit matching the block-size is reasonable, because anyway if the gap is smaller than that, you are most probably going to fetch the whole block anyway.

If we leave this a bit open, for instance by allowing users to set `0` or something small, then we would be in the same situation like we were when we started doing this whole `{min,max}` optimization (i.e., `libsonata` triggering millions of IO transactions). Therefore, I'd prefer if we could set a magic number limit, like 1 GPFS block, to at least limit how low you can set this parameter.

File: tests/test_report_reader.cpp
Comment (sergiorg-hpc) [reply tests/test_report_reader.cpp:92]: Perfect, will do right away, thank you so much.

Comment (sergiorg-hpc) [reply tests/test_report_reader.cpp:92]: @mgeplf I have rebased `master` and fix the test to match the correct output. One question, do you think it is worth it to also add the same test in Python? I have tried to make them consistent in this PR, but it is your call, of course.

File: include/bbp/sonata/report_reader.h
Comment (mgeplf) [reply include/bbp/sonata/report_reader.h:187]: Cool, glad we have this.

File: src/report_reader.cpp
Comment (mgeplf) [reply src/report_reader.cpp:85]: I'd just drop the templeting:
```
void emplace_ids(NodeID& key, NodeID node_id, ElementID /* element_id */) {
    key = node_id;
}

void emplace_ids(CompartmentID& key, NodeID node_id, ElementID element_id) {
    key[0] = node_id;
    key[1] = element_id;
}
```

File: src/report_reader.cpp
Comment (mgeplf) [reply src/report_reader.cpp:321]: I'm really hoping no-one ever tweaks the parameter in the first place, but that's an aside.

I'm just wondering if we're making it too GPFS centric; but I guess the same value makes sense on an NVME node?

File: tests/test_report_reader.cpp
Comment (mgeplf) [reply tests/test_report_reader.cpp:92]: > do you think it is worth it to also add the same test in Python?

Yeah, good point - I spaced on that, didn't want to block you, but wanted you to have something to test.

We don't have to do that in this PR, though.

File: src/report_reader.cpp
Comment (matz-e) [reply src/report_reader.cpp:321]: I think it makes sense for other systems, still. Even on NVMEs, we might want to have larger reads for lower latency…

**Review by mgeplf — DISMISSED** [review]
Awesome, and thanks for adding the python tests, it's appreciated.

File: src/report_reader.cpp
Comment (sergiorg-hpc) [reply src/report_reader.cpp]: Thank you so much, that actually makes sense. I have also added `inline` to your suggestion and verified that the performance is the same (or slightly better even):

```
   GIDs |   Elapsed |      Open |    Select |       Get
      1 |  1.673048 |  1.051501 |  0.619802 |  0.001734
     10 |  1.376986 |  0.721323 |  0.628078 |  0.027575
    100 |  4.257407 |  0.710185 |  0.638011 |  2.909204
   1000 |  7.380188 |  0.777585 |  0.625556 |  5.977037
  10000 |  7.389488 |  0.770053 |  0.642935 |  5.976491
 100000 |  7.823633 |  0.766917 |  0.633076 |  6.423629
1000000 | 11.833009 |  0.752722 |  0.698966 | 10.381311
4234929 | 20.914003 |  0.757421 |  0.268733 | 19.887840
```

File: src/report_reader.cpp
Comment (sergiorg-hpc) [reply src/report_reader.cpp:316]: As @matz-e says, it shouldn't matter because here we would just force to use the `{min,max}` optimization as much as possible. For the NVMe drives, the higher the requests also the better to reduce the latency.

File: tests/test_report_reader.cpp
Comment (sergiorg-hpc) [reply tests/test_report_reader.cpp:92]: Added! Thanks to @olupton for the tip about using `dtype=np.float32` to compare the values correctly.

@mgeplf We might consider to create a separate PR at some point to double-check the other tests that use floating point values, as right now they do approximate comparisons with [`np.testing.assert_allclose()`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html). More importantly, we should also add data tests for the `ElementReportReader`.

**Comment by sergiorg-hpc** [comment]
> Awesome, and thanks for adding the python tests, it's appreciated.

Thank you, @mgeplf.

By the way, I forgot to add a small thing in the Python tests, can you approve once again, please?

**Review by mgeplf — APPROVED** [review]

**Pull request merged**
